src/lightning/fabric/strategies/fsdp.py

```diff
@@ -506,20 +507,37 @@ def load_checkpoint(
             return metadata
 
         if _is_full_checkpoint(path):
-            if optimizers:
-                rank_zero_warn(
-                    "Loading a full-state checkpoint into FSDP currently only supports loading the model weights."
-                    " The optimizer state won't be reloaded."
-                )
-
             # This is inefficient, as multiple copies of the checkpoint are held in CPU memory at once.
             # There is currently no other way because `summon_full_params` does not support write-back from rank 0 only.
             checkpoint = torch.load(path, map_location="cpu")
             with FSDP.summon_full_params(module, writeback=True, rank0_only=False):
                 module.load_state_dict(checkpoint.pop(module_key), strict=strict)
 
+            # Load optimizer states
+            for optim_key, optim in optimizers.items():
+                # rank0_only should be false because we need to load the optimizer state on all ranks
+                with _get_full_state_dict_context(module, rank0_only=False):
+                    temp_state_dict = checkpoint[optim_key]
+
+                    # Handling the case where the optimizer state is saved from a normal optimizer
+                    if isinstance(list(checkpoint[optim_key]["state"].keys())[0], int):
+                        temp_state_dict = FSDP.rekey_optim_state_dict(
+                            temp_state_dict, OptimStateKeyType.PARAM_NAME, module
+                        )
```

Comment:
Any chance you could help me understand this a bit better? I've looked at this function `FSDP.rekey_optim_state_dict` in the past but never understood why this is needed. What do we consider a "normal optimizer" here. When we're loading from a full-state single file, isn't the state already "normal"?

Reply:
When saving from a "normal" optimizer, the state dictionary is indexed by `id` instead of `param_name`. However, for some reason (possibly due to shards where each device contains a different portion of the state), the `optim_state_dict_to_load` expects the state dictionary to be indexed by `param_name`.

Some reference [here](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.rekey_optim_state_dict):
> Re-keys the optimizer state dict optim_state_dict to use the key type optim_state_key_type. This can be used to achieve compatibility between optimizer state dicts from models with FSDP instances and ones without.

Reply:
When saving from a "normal" optimizer, the state dictionary is indexed by `id` instead of `param_name`. However, for some reason (possibly due to shards where each device contains a different portion of the state), the `optim_state_dict_to_load` expects the state dictionary to be indexed by `param_name`.

Some reference [here](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.rekey_optim_state_dict):
> Re-keys the optimizer state dict optim_state_dict to use the key type optim_state_key_type. This can be used to achieve compatibility between optimizer state dicts from models with FSDP instances and ones without.

Reply:
Got it, thanks!